src/rulesets/LinearAlgebra/structured.jl

```diff
@@ -47,6 +47,225 @@ end
 _symmetric_back(ΔΩ) = UpperTriangular(ΔΩ) + LowerTriangular(ΔΩ)' - Diagonal(ΔΩ)
 _symmetric_back(ΔΩ::Union{Diagonal,UpperTriangular}) = ΔΩ
 
+#####
+##### `Symmetric{<:Real}`/`Hermitian` eigendecomposition
+#####
+
+function frule((_, ΔA), ::typeof(eigen), A::LinearAlgebra.RealHermSymComplexHerm)
+    F = eigen(A)
+    ∂F = Thunk() do
+        λ, U = F
+        M = U' * ΔA * U
+        ∂λ = real(diag(M)) # if ΔA is Hermitian, so is U′ΔAU, so its diagonal is real
+        # K is skew-hermitian with zero diag
+        K = M ./ _nonzero.(λ' .- λ)
+        _setdiag!(K, Zero())
+        ∂U = U * K
+        return Composite{typeof(F)}(values = ∂λ, vectors = ∂U)
+    end
+    return F, ∂F
+end
+
+function rrule(::typeof(eigen), A::LinearAlgebra.RealHermSymComplexHerm)
+    F = eigen(A)
+    function eigen_pullback(ΔF)
+        ∂A = Thunk() do
+            λ, U = F
+            ∂λ, ∂U = ΔF.values, ΔF.vectors
+            # K is skew-hermitian
+            K = U' * ∂U
+            # unstable for degenerate matrices
+            U′∂AU = K ./ (_nonzero.(λ' .- λ))
+            setdiag!(U'∂AU, ∂λ)
+            return _symherm(A, U * U′∂AU * U')
+        end
+        return NO_FIELDS, ∂A
+    end
+    return F, eigen_pullback
+end
+
+function frule((_, ΔA), ::typeof(eigvals), A::LinearAlgebra.RealHermSymComplexHerm)
+    λ, U = eigen(A)
+    return λ, @thunk real(diag(U' * ΔA * U))
+end
+
+function rrule(::typeof(eigvals), A::LinearAlgebra.RealHermSymComplexHerm)
+    F, back = rrule(eigen, A)
+    λ, U = F
+    function eigvals_pullback(Δλ)
+        ∂A = Thunk() do
+            ∂F = Composite{typeof(F)}(values = Δλ)
+            return unthunk(back(∂F)[2])
+        end
+        return NO_FIELDS, ∂A
+    end
+    return λ, eigvals_pullback
+end
+
+_nonzero(x) = abs(x) > eps(eltype(x)) ? x : eps(eltype(x)) * sign(x)
+
+_setdiag!(A, d) = (A[diagind(A)] = d)
+_setdiag!(A, d::AbstractZero) = (A[diagind(A)] .= 0)
+_setdiag!(A::AbstractZero, d) = Diagonal(d)
+
+_pureimag(x) = x - real(x)
+
+function _realifydiag!(A)
+    for i in diagind(A)
+        @inbounds A[i] = real(A[i])
+    end
+    return A
+end
+_realifydiag!(A::LinearAlgebra.RealHermSym) = A
+
+_realifydiag(A) = A - _pureimag(Diagonal(A))
+_realifydiag(A::LinearAlgebra.RealHermSym) = A
+
+# constrain B to have same Symmetric/Hermitian type as A
+_symherm(A::Symmetric, B) = Symmetric(B, Symbol(A.uplo))
+_symherm(A::Hermitian, B) = Hermitian(B, Symbol(A.uplo))
+_symherm(f, B::AbstractMatrix{<:Real}, uplo) = Symmetric(B, uplo)
+_symherm(f, B::AbstractMatrix{<:Complex}, uplo) = Hermitian(B, uplo)
```

Comment:
Few thoughts 
- I wonder if this 3-arg version (where the first arg is not used) would be clearer as just another 2-arg version
`_symherm(B::AbstractMatrix{<:Complex}, uplo::Union{Char, Symbol})`
- I wonder if the mutating version should be a different function (since it does a subtly different thing, rather than the same thing but in-place)
- If we have `_symherm!`, then we want the mutated arg to be first, i.e. `_symherm!(B, A)` since it's `B` that (potentially) get's mutated. In which case we'd want to change the order for all methods, to be e.g. 
  ```julia
  # constrain B to have same Symmetric/Hermitian type as A
  _symherm(B::Symmetric, A) = Symmetric(B, Symbol(A.uplo))
  ```
- I wonder if we only actually need the `uplo` version? and then just call `_symherm(B, A.uplo)` instead of `_symherm(B, A)`? Or is getting `B::Hermition` if `A isa Hermitian` important (even if `eltype(B) <: Real`)?

Reply:
These are great points, thanks! Yeah, this gets to the funniness of base. For one of these functions `f` (e.g. `exp`), then we have `y = f(x::Hermitian{<:Real})::Symmetric{<:Real}`, and `y.uplo` is always `'U'`, even if `x.uplo` is `'L'`,
The exception is the (not yet implemented) type-unstable functions, for which we have a different set of rules, hence `f` as an argument for future dispatch:
```julia
_symherm(f, B::AbstractMatrix{<:Real}, uplo) = Symmetric(B, uplo)
_symherm(f, B::AbstractMatrix{<:Complex}, uplo) = Hermitian(B, uplo)
```
 However, we want outputs of pushforwards and pullbacks to exactly match the types and have the same `uplo` as their corresponding primals, so we also have
```julia
_symherm(A::Symmetric, B) = Symmetric(B, Symbol(A.uplo))
_symherm(A::Hermitian, B) = Hermitian(B, Symbol(A.uplo))
```
And lastly, for some reason the functions in base zero the imaginary component of the `.data` field in-place, hence `_symherm!`.

The point is well-taken though, and I can both clean up the signatures and document some of this logic.